server/pulp/server/lazy/url.py

```diff
@@ -0,0 +1,564 @@
+        :param pem: A PEM encoded key.
+        :type pem: str
+        :return: The loaded key.
+        :rtype: RSA.RSA
+        """
+        if path:
+            with open(path) as fp:
+                pem = fp.read()
+        bfr = BIO.MemoryBuffer(pem)
+        if 'PRIVATE' in pem:
+            key = RSA.load_key_bio(bfr)
+        else:
+            key = RSA.load_pub_key_bio(bfr)
+        return key
+
+
+class URL(object):
+    """
+    An URL object that supports signing.
+
+    :ivar content: The actual URL.
+    :type content: ParseResult
+    """
+
+    POLICY = 'policy'
+    SIGNATURE = 'signature'
+
+    def __init__(self, content):
+        """
+        :param content: The *actual* URL key.
+        :type content: str
+        """
+        self.content = urlparse(content)
+
+    @property
+    def scheme(self):
+        """
+        :return: The *scheme* component of the URL.
+        :rtype: str
+        """
+        return self.content.scheme
+
+    @property
+    def netloc(self):
+        """
+        :return: The *host:port* component of the URL.
+        :rtype: str
+        """
+        return self.content.netloc
+
+    @property
+    def path(self):
+        """
+        :return: The *path* component of the URL.
+        :rtype: str
+        """
+        return self.content.path
+
+    @property
+    def params(self):
+        """
+        :return: The *params* component of the URL.
+        :rtype: str
+        """
+        return self.content.params
+
+    @property
+    def query(self):
+        """
+        :return: The *query* component of the URL.
+        :rtype: str
+        """
+        return self.content.query
+
+    @property
+    def uri(self):
+        """
+        :return: The *uri* specified by the URL.
+        :rtype: str
+        """
+        path = self.path
+        params = self.params
+        query = self.query
+        if params:
+            params = ';{p}'.format(p=params)
+        if query:
+            query = '?{q}'.format(q=query)
+        uri = [
+            path,
+            params,
+            query
+        ]
+        return ''.join(uri)
+
+    def sign(self, key, expiration=10):
+        """
+        Sign the URL.
+        Has the format of: <url>?policy=<policy>;signature=<signature>.
+        The policy is: {uri: <URI>, expiration: <expiration>}
+        The uri is the path?query in the original URL.
+        The expiration is: seconds since epoch.
+        The signature is RSA signature of the SHA256 digest of the
+        json/base64 encoded policy.
+
+        :param key: A public RSA key.
+        :type key: RSA.RSA
+        :param expiration: The signature expiration in seconds.
+        :type expiration: int
+        :return: The signed URL.
+        :rtype: SignedURL
+        """
+        expiration = int(time() + expiration)
+        policy = Policy(self.uri, expiration)
+        policy, signature = policy.sign(key)
+        query = Query.decode(self.content.query)
```

Comment:
Any particular reason to get the query from inside `self.content` again rather than `self.query`?